### PR TITLE
feat(core): freeze RFC-64 semantic record codec

### PR DESCRIPTION
**Pull request by branarakic**
## Outcome

Adds a pure, backend-neutral codec for the seven RFC-64 semantic control records. It projects validated records into exact typed RDF rows and decodes rows only when their authenticated type, scope, graph, subject, cardinality, datatypes, lexical forms, nulls, and resource bounds all match.

The codec derives all semantic graph and subject addresses internally from the previous stack item. Callers cannot select those addresses.

## User impact

There is no user-visible runtime behavior change yet: this increment is deliberately dormant and does not add storage writes, SPARQL queries, CAS operations, or RFC-64 activation.

It creates the deterministic boundary needed for the next shippable slice. Oxigraph and Blazegraph adapters can now be tested against one exact semantic representation before that state is wired into atomic sync apply and recovery.

## Before

```mermaid
sequenceDiagram
    participant Apply as Future RFC-64 apply lane
    participant Backend as Oxigraph or Blazegraph
    Apply->>Backend: Adapter-specific record fields
    Backend-->>Apply: Backend-shaped RDF terms
    Note over Apply,Backend: Predicates, nulls, lists, and cardinality are not frozen
```

## After

```mermaid
sequenceDiagram
    participant Apply as Future RFC-64 apply lane
    participant Codec as Pure core semantic codec
    participant Backend as Oxigraph or Blazegraph
    Apply->>Codec: Typed record plus authenticated scope
    Codec->>Codec: Derive fixed graph and subject
    Codec-->>Backend: Closed typed RDF rows
    Backend-->>Codec: Typed rows in any order
    Codec->>Codec: Enforce scope, cardinality, terms, nulls, and bounds
    Codec-->>Apply: Exact decoded record
```

## What is frozen

- Seven exact semantic record shapes.
- Closed predicate and RDF-object registry.
- Explicit `urn:dkg:sync:null` representation for every nullable field.
- Canonical JCS pending-digest list with a 64-entry bound.
- Exact per-record row counts and order-independent inverse decoding.
- Fixed graph and subject derivation from authenticated scope.
- Canonical integer, digest, address, datetime, and datatype validation.
- 64 KiB cumulative typed-row response bound.
- Snapshotting against post-validation caller mutation.

## Stack

- Depends on #2382.
- Depends transitively on #2377.
- Normative companion: OriginTrail/dkgv10-spec#151.

## Validation

- `pnpm --filter @origintrail-official/dkg-core build`
- Full core lane: 114 files / 1,754 tests passed.
- Focused codec/address/barrel lane: 3 files / 22 tests passed.
- Direct oxlint on all changed TypeScript files.
- `git diff --check`

### CODE REVIEW

File: packages/core/src/rfc64-semantic-records-v1.ts [hunk 1]
```diff
@@ -0,0 +1,954 @@
...
+  const value = snapshotClosed(
+    envelope.value,
+    fields.map((field) => field.key),
+    recordType,
+  );
+  validateRecordValue(recordType, value);
+  const immutableValue = recordType === 'SubgraphReconcileTargetGuardV1'
+    ? Object.freeze({
+        ...value,
+        pendingTargetCheckpointDigests: Object.freeze([
```
Comment (otReviewAgent) [inline packages/core/src/rfc64-semantic-records-v1.ts]: 🔴 Bug: The pending-digest snapshot re-reads untrusted data after validation

**What's wrong**
The code validates descriptor values, then snapshots the original array using spread. For a Proxy, those are separate observable reads. The spread can therefore copy invalid, duplicate, or even unbounded iterator values that were never validated, breaking the codec's validated-record guarantee and potentially enabling denial of service.

**Example**
A Proxy around `[validDigest]` can report `validDigest` from its `getOwnPropertyDescriptor` trap but return `'not-a-digest'` from its `get` trap. Validation passes, yet the returned frozen record contains `['not-a-digest']`, and projection serializes that invalid value.

**Suggested direction**
Have pending-list validation return an immutable snapshot built directly from the same descriptor values it validates, then place that snapshot into the record.

**For Agents**
In `snapshotRfc64SemanticRecordV1`/`assertPendingDigestList`, collect and validate descriptor values in one pass and freeze that collected list; never spread or iterate the original array afterward. Add an adversarial Proxy test proving the projected record retains only the values that were actually validated and round-trips successfully.

File: packages/core/test/rfc64-semantic-records-v1.test.ts [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+
+describe('RFC-64 semantic record RDF codec v1', () => {
+  it('round-trips all seven records independent of backend row order', () => {
+    for (const record of records) {
+      const rows = projectRfc64SemanticRecordStoreRowsV1(record);
+      const decoded = decodeRfc64SemanticRecordStoreRowsV1(
+        [...rows].reverse(),
+        coordinateFor(record),
+      );
+      expect(decoded.record, record.recordType).toEqual(record);
```
Comment (otReviewAgent) [inline packages/core/test/rfc64-semantic-records-v1.test.ts:132]: 🔴 Bug: The round-trip test is self-referential for six RDF record contracts

**What's wrong**
The encoder and decoder share the mapping under test, so their agreement does not verify that six record types use the required RDF contract. This gives false confidence around persisted semantic data and interoperability.

**Example**
If `AppliedContextGraphSealV1` accidentally swapped the `policyDigest` and `chainCoverageDigest` predicate mappings, projection and decoding would make the same mistake, so lines 130–131 would still pass even though persisted RDF no longer matches the external contract. The same weakness applies to selecting the wrong semantic graph address.

**Suggested direction**
Add externally defined golden vectors for every record type, especially their record-specific predicates and semantic graph/subject addresses.

**For Agents**
In `packages/core/test/rfc64-semantic-records-v1.test.ts`, add independent golden rendered-row fixtures for the other six record types. Assert exact subject, graph, predicate, lexical value, and datatype without deriving expected values through the decoder or shared production mappings.

File: packages/core/src/rfc64-semantic-records-v1.ts [hunk 3]
```diff
@@ -0,0 +1,954 @@
...
+const P = RFC64_SEMANTIC_PREDICATES_V1;
+const COMMON_SCOPE_FIELDS = [
+  { key: 'networkId', predicate: P.NETWORK_ID, kind: 'string' },
+  { key: 'contextGraphId', predicate: P.CONTEXT_GRAPH_ID, kind: 'string' },
+] as const satisfies readonly FieldSpec[];
+const SUBGRAPH_FIELD = {
+  key: 'subGraphName', predicate: P.SUBGRAPH_NAME, kind: 'nullable-string',
+} as const satisfies FieldSpec;
+
+const RECORD_FIELDS = Object.freeze({
```
Comment (otReviewAgent) [inline packages/core/src/rfc64-semantic-records-v1.ts]: 🟡 Issue: Make the record definition registry the actual source of truth

**What's wrong**
The implementation appears data-driven, but the table controls only field enumeration and RDF representation. The real schema remains duplicated among public interfaces, weak string keys, special-case validation, coordinate construction, address routing, and deep-copy branches. That makes future record evolution a multi-site edit with no compiler enforcement and leaves an almost-1k-line module whose abstractions do not contain the complexity they introduce.

**Example**
Changing a descriptor to `{ key: 'catalogHeadDigset', ... }` still satisfies `FieldSpec`. A statically valid `CurrentAuthorCatalogRefV1` would then be rejected at runtime because the closed-object schema silently diverged from its interface. Adding an eighth record similarly requires synchronized edits across multiple distant branch chains.

**Suggested direction**
Promote this into a typed per-record definition registry containing the fields, coordinate projection/keys, address derivation, and any exceptional validation or snapshot policy. Associate fields with domain scalar codecs rather than a loose `FieldKind` string—currently, for example, `nullable-integer` implicitly means `ChainIdV1` while `string` validation happens outside the field loop. Derive the record-type union and generic codec flow from those definitions, then split the public models/definitions from the reusable codec engine. This should delete most condition chains and casts instead of moving them around.

**For Agents**
Refactor the definitions around `RECORD_FIELDS`. Preserve all seven wire shapes, row order, addresses, error categories, and immutable snapshots. Make invalid field keys fail typechecking, remove the repeated record-type dispatch where practical, and keep the existing round-trip tests green; add a compile-time assertion or fixture proving a descriptor cannot name a field absent from its value type.

File: packages/core/src/rfc64-semantic-records-v1.ts [hunk 4]
```diff
@@ -0,0 +1,954 @@
...
+      | 'SubgraphMutationGuardV1'
+      | 'SubgraphReconcileTargetGuardV1' }
+    & Rfc64SubgraphSemanticScopeFieldsV1)
+  | ({ readonly recordType:
+      | 'ContextGraphMutationGuardV1'
+      | 'AppliedSubgraphSetRefV1'
+      | 'AppliedContextGraphSealV1' }
+    & Rfc64SemanticScopeFieldsV1);
+
+export type Rfc64SemanticStoreObjectV1 = CanonicalAuthorSealStoreObjectV1;
```
Comment (otReviewAgent) [inline packages/core/src/rfc64-semantic-records-v1.ts]: 🟡 Issue: Do not make generic RDF-row primitives belong to the author-seal codec

**What's wrong**
This new semantic codec depends on another feature codec for general-purpose types, then duplicates much of that codec's machinery because the implementation is private or author-seal-specific. The ownership direction is backwards and creates two implementations of the same boundary invariants, making both modules harder to evolve consistently.

**Example**
Any future change to the shared typed-row shape must be coordinated between the author-seal implementation and this separate implementation. Supporting the semantic digest-list datatype already required a second renderer because the author-seal renderer hard-codes its own datatype set.

**Suggested direction**
Create a neutral typed RDF-row module that owns the row/object types, safe structural snapshotting, constructors, rendering primitives, equality, and byte accounting, with codec-specific datatype admission layered above it. Move the canonical UTC-millisecond contract to the date-time/scalar layer as well. Re-export aliases if compatibility requires the existing names.

**For Agents**
Extract the backend-neutral RDF store object/row contract and its structural helpers from `canonical-graph-scoped-author-seal.ts` into a neutral core module, allowing codec-specific datatype policy and error mapping. Likewise move the canonical UTC-millisecond type/validator to the existing date-time boundary or another neutral scalar module. Update both codecs to consume those primitives while preserving their current public aliases, validation strictness, and rendered output; run both author-seal and RFC-64 semantic-record suites.

File: packages/core/test/rfc64-semantic-records-v1.test.ts [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+      .toEqual(record);
+  });
+
+  it('rejects missing, duplicate, and unknown predicates', () => {
+    const record = records[1];
+    const rows = [...projectRfc64SemanticRecordStoreRowsV1(record)];
+    expect(() => decodeRfc64SemanticRecordStoreRowsV1(
+      rows.slice(1),
+      coordinateFor(record),
+    )).toThrow(/exactly 8 rows/u);
```
Comment (otReviewAgent) [inline packages/core/test/rfc64-semantic-records-v1.test.ts:325]: 🟡 Issue: Rejection tests do not verify the public error-code contract

**What's wrong**
The tests verify error prose but not the machine-readable classification introduced by this codec. Incorrect or collapsed error codes could therefore ship without detection.

**Example**
Changing the noncanonical-generation failure from `rfc64-semantic-scalar` to `rfc64-semantic-row-term` while keeping the message unchanged would leave the current test green, but callers branching on `error.code` would observe a contract regression.

**Suggested direction**
Assert the structured error type and code for each distinct failure category instead of checking only human-readable message fragments.

**For Agents**
Add representative negative cases in `packages/core/test/rfc64-semantic-records-v1.test.ts` for each public error category, catch the error, and assert it is `Rfc64SemanticRecordErrorV1` with the expected `code`. Include malformed coordinates and malformed row structures.

File: packages/core/src/rfc64-semantic-records-v1.ts (hunk 1 above)
Comment (branarakic) [reply packages/core/src/rfc64-semantic-records-v1.ts]: Fixed in f520ab458. Pending digest validation now returns the frozen descriptor-value snapshot it validates; no spread or indexed read of the source follows validation. An adversarial Array Proxy test proves indexed getters are never invoked and the validated record projects/decodes successfully.

File: packages/core/test/rfc64-semantic-records-v1.test.ts (hunk 2 above)
Comment (branarakic) [reply packages/core/test/rfc64-semantic-records-v1.test.ts:132]: Fixed in f520ab458. Added independent exact RDF golden vectors for the other six record types, pinning literal predicate IRIs, graph/subject routes, lexical values, datatypes, and row order without using the production decoder/mapping for expectations.

File: packages/core/src/rfc64-semantic-records-v1.ts (hunk 3 above)
Comment (branarakic) [reply packages/core/src/rfc64-semantic-records-v1.ts]: Fixed in f520ab458. The typed  registry now owns each record's statically checked field keys, domain scalar codecs, coordinate keys, address derivation, and exceptional validation. The generic projection/snapshot/decode paths consume that registry, and the public models are split into . A misspelled field key no longer typechecks through .

File: packages/core/src/rfc64-semantic-records-v1.ts (hunk 4 above)
Comment (branarakic) [reply packages/core/src/rfc64-semantic-records-v1.ts]: Fixed in f520ab458. Added neutral  for row/object types, structural snapshotting, constructors, rendering, equality, and byte accounting. Both author-seal and semantic codecs consume it with codec-specific datatype/error adapters. The canonical UTC-millisecond brand/validator now lives in  and remains re-exported compatibly.

File: packages/core/test/rfc64-semantic-records-v1.test.ts (hunk 5 above)
Comment (branarakic) [reply packages/core/test/rfc64-semantic-records-v1.test.ts:325]: Fixed in f520ab458. The semantic-record suite now asserts  for all seven public categories, including malformed coordinates, malformed/adorned row structures, cardinality, term, scalar, schema, and byte-limit failures. Full core validation: 114 files, 1,761 tests passed.

Comment (otReviewAgent) [reply packages/core/test/rfc64-semantic-records-v1.test.ts:325]: 🔴 Bug: The wrong-datatype test never changes a datatype

**What's wrong**
The test title claims to verify rejection of wrong RDF datatypes, but its integer, digest, datetime, and list cases all use the correct datatype and only corrupt the lexical value. A regression that stopped enforcing exact datatype equality would therefore pass this test.

**Example**
Change a valid generation row to `{ kind: 'literal', value: '13', datatypeIri: 'http://www.w3.org/2001/XMLSchema#string' }`. The decoder should reject it as `rfc64-semantic-row-term`; removing the datatype comparison would currently leave this suite green.

**Suggested direction**
Add a case that substitutes one valid but incorrect datatype IRI and asserts the public error code, not only an error-message substring.

**For Agents**
In `packages/core/test/rfc64-semantic-records-v1.test.ts`, mutate a projected literal to a different valid datatype while preserving its lexical value. Assert the decoder throws `Rfc64SemanticRecordErrorV1` with code `rfc64-semantic-row-term`.

File: packages/core/src/rfc64-semantic-records-v1.ts (hunk 3 above)
Comment (otReviewAgent) [reply packages/core/src/rfc64-semantic-records-v1.ts]: 🟡 Issue: Make the schema registry exhaustive instead of erasing its invariants

**What's wrong**
The central registry looks strongly typed, but it only checks that each supplied field is valid—not that the definition is complete or unique. The implementation then discards the remaining type relationship to make the generic pipeline work. That makes the declared record interfaces and runtime schemas capable of silently drifting apart, defeating the main maintainability benefit of a schema-driven design.

**Example**
Removing `appliedAt` from `AppliedSubgraphSealV1.fields` still type-checks. The codec would then accept and return a value without the interface's required `appliedAt`, because the final result is cast back to `Rfc64SemanticRecordV1`. Duplicating a field key is likewise permitted by the type.

**Suggested direction**
Use an exhaustive mapped field object or a record factory that accepts every model key exactly once, normalizes it to ordered fields, and exposes type-correlated operations. This should concentrate any unavoidable cast inside the factory rather than spreading `unknown`/`never` casts through projection and decoding.

**For Agents**
Refactor `RECORD_DEFINITIONS` in `rfc64-semantic-records-v1.ts` into an exhaustive field map keyed by every `keyof T`, preserving canonical ordering explicitly if needed. Validate unique predicates and coordinate keys when defining a record, and keep record-type/value correlation inside the factory or dispatcher. Preserve the seven existing row layouts and round trips; add a compile-time fixture or definition-construction check proving omitted and duplicate fields cannot enter the registry.

File: packages/core/test/rfc64-semantic-records-v1.test.ts (hunk 5 above)
Comment (otReviewAgent) [reply packages/core/test/rfc64-semantic-records-v1.test.ts:325]: 🟡 Issue: Resource-limit tests omit the contract boundaries

**What's wrong**
Both new resource limits are tested only with values comfortably below and above the boundary. Off-by-one regressions, and character-counting in place of UTF-8 byte-counting, could reject valid protocol records without failing this suite.

**Example**
Changing the pending-list check from `length > 64` to `length >= 64` would still pass all current tests while rejecting a valid 64-entry history. Likewise, changing the byte check from `>` to `>=` would leave the current oversized test green while rejecting an exact-limit response.

**Suggested direction**
Add acceptance tests at each exact maximum and rejection tests immediately above it, including a multibyte UTF-8 case for the byte limit.

**For Agents**
Extend the resource-bound test in `packages/core/test/rfc64-semantic-records-v1.test.ts`. Prove that exactly 64 unique digests are accepted and 65 are rejected. For row bytes, construct exact-limit and one-byte-over inputs—preferably including a multibyte literal—and assert the mapped public error code.

File: packages/core/src/typed-rdf-store-row-v1.ts [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+export function typedRdfLiteralV1(value: string, datatypeIri: string): TypedRdfStoreObjectV1 {
+  if (!isSafeIri(datatypeIri)) throw new TypedRdfStoreRowErrorV1('row-term', 'literal datatype must be one bare safe IRI');
+  return Object.freeze({ kind: 'literal' as const, value, datatypeIri });
+}
+
+export function snapshotTypedRdfStoreRowV1(input: unknown): TypedRdfStoreRowV1 {
+  const row = snapshotClosed(input, ['graphIri', 'object', 'predicateIri', 'subjectIri'], 'typed RDF store row');
+  for (const key of ['subjectIri', 'predicateIri', 'graphIri'] as const) {
+    if (typeof row[key] !== 'string' || !isSafeIri(row[key])) {
```
Comment (otReviewAgent) [inline packages/core/src/typed-rdf-store-row-v1.ts]: 🔴 Bug: Non-string IRI fields are now misclassified as term errors

**What's wrong**
The shared validator combines a schema check with an RDF-term check and reports both as `row-term`. Because the existing author-seal API exposes and maps these error codes, this changes its behavior for malformed adapter rows and can break callers that distinguish malformed row structure from invalid RDF terms.

**Example**
Calling `renderCanonicalAuthorSealStoreRowV1({ ...validRow, subjectIri: 42 as any })` previously produced `canonical-seal-row-schema`; it now produces `canonical-seal-row-term`. An unsafe string such as `"urn:has space"` should remain a term error.

**Suggested direction**
Check that each row IRI field is a string first and throw `row-schema` when it is not; only apply `isSafeIri` afterward and classify that failure as `row-term`.

**For Agents**
In `snapshotTypedRdfStoreRowV1`, separate field-type validation from safe-IRI validation. Preserve `row-term` for unsafe strings, but return `row-schema` for non-string fields. Prove that the canonical author-seal wrapper retains its existing public error codes for both cases.

File: packages/core/src/canonical-graph-scoped-author-seal.ts [hunk 7]
```diff
@@ -580,46 +565,26 @@ export function projectCanonicalGraphScopedAuthorSealStoreRowsV1(
...
+      XSD_STRING_IRI,
+      XSD_INTEGER_IRI,
+      XSD_HEX_BINARY_IRI,
+      XSD_DATE_TIME_IRI,
+    ]));
+  } catch (cause) {
+    const code = cause instanceof TypedRdfStoreRowErrorV1 && cause.code === 'row-schema'
+      ? 'canonical-seal-row-schema'
+      : 'canonical-seal-row-term';
+    const message = cause instanceof TypedRdfStoreRowErrorV1
```
Comment (otReviewAgent) [inline packages/core/src/canonical-graph-scoped-author-seal.ts]: 🟡 Issue: Do not translate shared errors by rewriting their prose

**What's wrong**
The new shared RDF-row abstraction leaks its exact English messages into the author-seal layer. The adapter preserves domain wording by performing duplicated substring replacements, creating a hidden coupling where harmless edits to shared diagnostics alter another module's public error behavior without a type or compiler failure.

**Example**
If the shared helper changes `typed RDF literal has an invalid field set` to `typed RDF literal fields are invalid`, this adapter still compiles but silently stops translating the message and leaks the shared helper's terminology through the canonical-seal API.

**Suggested direction**
Give typed-row failures stable structured reasons or context fields and translate those once at the domain boundary. At minimum, centralize the mapping in one helper rather than duplicating message substitutions in render and decode paths.

**For Agents**
Inspect the typed-row error model and both author-seal catch blocks. Add a stable structured reason/context to `TypedRdfStoreRowErrorV1`, or a single canonical adapter that switches on structured data. Preserve the existing canonical-seal error codes and externally required messages; verify representative schema, term, and cardinality failures through the author-seal entry points.

File: packages/core/test/rfc64-semantic-records-v1.test.ts (hunk 2 above)
Comment (otReviewAgent) [reply packages/core/test/rfc64-semantic-records-v1.test.ts:132]: 🔴 Bug: The “wrong datatypes” test never changes a datatype

**What's wrong**
This test only verifies noncanonical lexical forms. It gives false confidence that the decoder's datatype binding is protected, even though weakening `exactLiteralValue` to accept the same lexical value under `xsd:string` would not fail any test.

**Example**
Change the generation row to `{ kind: 'literal', value: '13', datatypeIri: 'http://www.w3.org/2001/XMLSchema#string' }` and assert decoding fails with `rfc64-semantic-row-term`. A regression that ignored datatype IRIs would currently pass this suite.

**Suggested direction**
Add at least one genuine wrong-datatype decoder case, ideally covering a scalar and the custom digest-list datatype.

**For Agents**
In `rfc64-semantic-records-v1.test.ts`, add decoder cases that preserve valid lexical values while substituting another allowed datatype. Prove each field codec rejects the wrong RDF datatype/object form.

File: packages/core/src/index.ts [hunk 8]
```diff
@@ -71,6 +71,8 @@ export * from './author-catalog-objects.js';
 export * from './author-catalog-directory.js';
 export * from './swm-author-inventory-v1.js';
 export * from './rfc64-semantic-addresses-v1.js';
+export * from './rfc64-semantic-records-v1.js';
+export * from './typed-rdf-store-row-v1.js';
```
Comment (otReviewAgent) [inline packages/core/src/index.ts:76]: 🟡 Issue: The newly public typed-RDF API has no barrel verification

**What's wrong**
The codec wrappers indirectly exercise parts of the utility, but they do not verify that its newly advertised root exports exist or preserve their own public contracts.

**Example**
Removing the `export * from './typed-rdf-store-row-v1.js'` line would leave the current tests green while breaking consumers importing `snapshotTypedRdfStoreRowV1` from `@origintrail-official/dkg-core`.

**Suggested direction**
Add a public-surface test for the typed-RDF exports, plus focused direct assertions for the generic error codes and validation behavior consumers receive.

**For Agents**
Extend the core barrel-export test to import representative typed-RDF functions and `TypedRdfStoreRowErrorV1` from `../src/index.js`. Add direct helper tests where wrapper tests translate or hide the generic API's own error contract.

Comment (branarakic) [reply packages/core/src/index.ts:76]: Addressed in a222668de. The package-barrel test now imports representative typed-RDF functions and TypedRdfStoreRowErrorV1 through ../src/index.js, verifies valid snapshot/render behavior, and directly pins row-schema versus row-term failures.

File: packages/core/src/typed-rdf-store-row-v1.ts (hunk 6 above)
Comment (branarakic) [reply packages/core/src/typed-rdf-store-row-v1.ts]: Addressed on the current head: non-string IRI fields are classified as row-schema before safe-IRI validation; unsafe strings remain row-term, with wrapper regressions preserving canonical-seal public codes.

File: packages/core/src/canonical-graph-scoped-author-seal.ts (hunk 7 above)
Comment (branarakic) [reply packages/core/src/canonical-graph-scoped-author-seal.ts]: Addressed on the current head: TypedRdfStoreRowErrorV1 now exposes structured reasons and the author-seal boundary translates them through one typed adapter rather than rewriting diagnostic prose.

File: packages/core/src/rfc64-semantic-record-model-v1.ts [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+  PENDING_TARGET_CHECKPOINT_DIGESTS: `${DKG_ONTOLOGY}pendingTargetCheckpointDigests`,
+  SUBGRAPH_INDEX_ERA: `${DKG_ONTOLOGY}subgraphIndexEra`,
+  SUBGRAPH_INDEX_VERSION: `${DKG_ONTOLOGY}subgraphIndexVersion`,
+  SUBGRAPH_COUNT: `${DKG_ONTOLOGY}subgraphCount`,
+  APPLIED_DIRECTORY_ROOT_DIGEST: `${DKG_ONTOLOGY}appliedDirectoryRootDigest`,
+  POLICY_DIGEST: `${DKG_ONTOLOGY}policyDigest`,
+  CHAIN_COVERAGE_DIGEST: `${DKG_ONTOLOGY}chainCoverageDigest`,
+} as const);
+
+interface Rfc64SemanticScopeFieldsV1 {
```
Comment (otReviewAgent) [inline packages/core/src/rfc64-semantic-record-model-v1.ts]: 🟡 Issue: Reuse the canonical scope and rendered-row contracts instead of shadowing them

**What's wrong**
The new model recreates shared contracts that already exist in the modules responsible for semantic addresses and typed RDF rows. These parallel structural types add schema owners and drift risk without providing a distinct semantic invariant.

**Example**
A change to the canonical subgraph scope or rendered-row contract now requires maintainers to discover and update a second structurally identical declaration in this model, even though address derivation and row rendering already own those concepts elsewhere.

**Suggested direction**
Have record interfaces extend `Rfc64SemanticScopeV1`, `Rfc64SubgraphSemanticScopeV1`, or `Rfc64AuthorSemanticScopeV1` as appropriate, and make the semantic rendered-row type an alias of the generic rendered-row model. This leaves one owner for each shared boundary.

**For Agents**
In `rfc64-semantic-record-model-v1.ts`, import and extend the canonical RFC-64 scope interfaces from `rfc64-semantic-addresses-v1.ts`, and alias or extend `RenderedRdfStoreRowV1` from `typed-rdf-store-row-v1.ts`. Preserve the exported semantic record shapes and verify with the core TypeScript build and barrel-export tests.

Comment (branarakic) [reply packages/core/src/rfc64-semantic-record-model-v1.ts]: Fixed in 0da9d76c3. Semantic record interfaces now extend the canonical RFC-64 semantic scope contracts, CurrentAuthorCatalogRefV1 uses the author scope directly, coordinates reuse the same scope types, and the rendered-row type is an alias of RenderedRdfStoreRowV1. Core build, 14 semantic-record tests, address tests, and lint pass.

File: packages/core/src/typed-rdf-store-row-v1.ts [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+  if (left.object.kind === 'named-node' && right.object.kind === 'named-node') {
+    return left.object.value === right.object.value;
+  }
+  return left.object.kind === 'literal'
+    && right.object.kind === 'literal'
+    && left.object.value === right.object.value
+    && left.object.datatypeIri === right.object.datatypeIri;
+}
+
+export function typedRdfStoreRowByteLengthV1(row: TypedRdfStoreRowV1): number {
```
Comment (otReviewAgent) [inline packages/core/src/typed-rdf-store-row-v1.ts:214]: 🟡 Issue: The byte-limit test does not verify UTF-8 byte accounting

**What's wrong**
The new response-size guard is security/resource-bound behavior, but its test data cannot distinguish byte counting from character counting. Consequently, a regression that undercounts non-ASCII RDF values would remain green and could admit substantially larger responses.

**Example**
A regression replacing UTF8.encode(value).byteLength with value.length would still pass the current ASCII test. For example, 'é'.repeat(MAX_RFC64_SEMANTIC_RECORD_RESPONSE_BYTES_V1 / 2) occupies 64 KiB in UTF-8 before the other row fields are counted, but only half that many JavaScript code units; decoding should reject it as oversized.

**Suggested direction**
Add a multibyte boundary case whose UTF-8 size exceeds the limit while its JavaScript string length remains below it. A direct assertion for typedRdfStoreRowByteLengthV1 would also pin the public helper's units.

**For Agents**
Add coverage near the semantic-record resource-bound test, or a focused typed-rdf-store-row-v1 test, using multibyte subject/object content. Prove both typedRdfStoreRowByteLengthV1's UTF-8 result and decodeRfc64SemanticRecordStoreRowsV1's 64 KiB rejection.

Comment (branarakic) [reply packages/core/src/typed-rdf-store-row-v1.ts:214]: Addressed in 72081be5a. The semantic-record resource-bound test now uses a multibyte literal whose JavaScript length is below 64 KiB while its UTF-8 row size exceeds the limit. It directly asserts typedRdfStoreRowByteLengthV1 reports bytes and that decodeRfc64SemanticRecordStoreRowsV1 rejects the response as oversized.

Comment (otReviewAgent) [reply packages/core/src/typed-rdf-store-row-v1.ts:214]: 🔴 Bug: The byte limit allocates the entire oversized value before rejecting it

**What's wrong**
The newly added response-size guard is intended to bound untrusted semantic rows, but its measurement routine fully UTF-8-encodes each string before checking the limit. A sufficiently large backend or caller-supplied literal can therefore cause a large temporary allocation or out-of-memory failure before the guard throws.

**Example**
A row containing a 500 MB literal reaches `UTF8.encode(row.object.value)` and can allocate hundreds of megabytes before line 156 reports `row-too-large`, potentially exhausting the process instead of safely rejecting the response.

**Suggested direction**
Perform bounded byte counting. A cheap UTF-16-length lower-bound check can reject obviously oversized strings first; only encode values small enough to fit the remaining configured budget, or use fixed-size `encodeInto` chunks.

**For Agents**
Update the byte-accounting path in `typed-rdf-store-row-v1.ts` to reject strings that cannot fit in the remaining budget before fully encoding them, while preserving exact UTF-8 accounting for values near the limit. Add boundary cases for ASCII and multibyte literals and prove a grossly oversized value takes the early-rejection path.

File: packages/core/src/canonical-graph-scoped-author-seal.ts (hunk 7 above)
Comment (otReviewAgent) [reply packages/core/src/canonical-graph-scoped-author-seal.ts]: 🟡 Issue: Do not translate shared errors by comparing helper prose

**What's wrong**
The new shared RDF abstraction is coupled back to the author-seal implementation through exact internal label strings. A harmless wording or helper-context refactor in the generic module can now change author-seal behavior without any type error. This replaces the previous duplication with a brittle cross-module protocol rather than a clean abstraction boundary.

**Example**
Renaming the helper label from `typed RDF literal` to `RDF literal` leaves generic validation and error codes unchanged, but silently stops the translation to `literal object has invalid fields` and breaks the author-seal contract.

**Suggested direction**
Give the shared layer stable semantic reason tags such as `invalid-literal-field-set` and `non-data-discriminator`, or inject a domain error mapper at the boundary. Human-readable labels should remain presentation details and never serve as dispatch keys.

**For Agents**
Inspect `typed-rdf-store-row-v1.ts` error reasons and `translateTypedAuthorSealRowFailure`. Preserve current generic and author-seal codes/messages, but replace label comparisons with stable semantic variants or an explicit caller-owned error policy. Existing malformed-row tests should continue to prove the legacy messages and codes.

File: packages/core/test/rfc64-semantic-records-v1.test.ts [hunk 11]
```diff
@@ -0,0 +1,636 @@
...
+      ...root,
+      value: {
+        ...root.value,
+        governanceChainId: CHAIN,
+        governanceContractAddress: null,
+      },
+    })).toThrow(/must both be null or both non-null/u);
+  });
+
+  it('enforces the pending-history and typed-response resource bounds', () => {
```
Comment (otReviewAgent) [inline packages/core/test/rfc64-semantic-records-v1.test.ts:456]: 🟡 Issue: Valid pending-history boundary values are not verified

**What's wrong**
The suite proves that two entries work and 65 fail, but not that the advertised inclusive maximum of 64 works or that an empty pending history is accepted. Off-by-one validation or an accidental nonempty requirement could therefore reject valid persisted guard records while all tests stay green.

**Example**
Round-trip SubgraphReconcileTargetGuardV1 once with pendingTargetCheckpointDigests: [] and once with 64 unique canonical digests. Both should project and decode successfully; 65 should remain rejected.

**Suggested direction**
Test both valid endpoints—an empty pending history and exactly 64 unique entries—alongside the existing 65-entry rejection.

**For Agents**
Extend the resource-bound test in rfc64-semantic-records-v1.test.ts with successful projection and decode cases at zero and exactly 64 pending digests, preserving uniqueness and canonical digest formatting.

Comment (branarakic) [reply packages/core/test/rfc64-semantic-records-v1.test.ts:456]: Addressed in 5a315e42e. The resource-bound suite now round-trips both valid endpoints—an empty pending history and exactly 64 unique canonical digests—while preserving the existing 65-entry and duplicate rejection checks. Focused core suite: 14/14 passed.

File: packages/core/src/typed-rdf-store-row-v1.ts (hunk 6 above)
Comment (otReviewAgent) [reply packages/core/src/typed-rdf-store-row-v1.ts]: 🟡 Issue: Do not use display labels as the cross-module error protocol

**What's wrong**
The shared RDF abstraction leaks its internal wording into a consumer's control flow. Benign message or helper-label refactors can therefore change author-seal behavior without a type error, making the abstraction brittle and coupling modules that should communicate through structured errors.

**Example**
Renaming the internal label `typed RDF literal` to `RDF literal` would compile successfully but silently bypass the author-seal translation to `literal object has invalid fields`.

**Suggested direction**
Give each reason a stable typed location, such as `location: 'row' | 'literal' | 'named-node'`, and a constrained property identifier where needed. Keep human-readable labels only in messages. This lets domain adapters translate exhaustively without depending on wording from the shared helper.

**For Agents**
Replace the free-form control fields in `typed-rdf-store-row-v1.ts` with stable typed locations/properties, then update `translateTypedAuthorSealRowFailure` to switch exhaustively on them. Preserve the existing generic error codes and author-seal-facing codes/messages; existing malformed-row tests should continue to prove that mapping.

**Comment by branarakic** [comment]
Superseded by merged integration PR #2432. The current head of this stacked PR is contained in testnet-canary.
